File: augur-engine/src/collect/gdelt.ts
import type { Signal } from '../config/types.js';

const GDELT_DOC_API = 'https://api.gdeltproject.org/api/v2/doc/doc';

/** Collect geopolitical signals from GDELT Cloud API */
export async function collectGdelt(): Promise<Signal> {
  // GDELT DOC API: latest articles mentioning key themes
  const params = new URLSearchParams({
    query: 'theme:ECON_BANKRUPTCY OR theme:ENV_CLIMATECHANGE OR theme:CRISISLEX_CRISISLEXREC OR theme:MILITARY',
    mode: 'ArtList',
    maxrecords: '20',
    format: 'json',
    sort: 'DateDesc',
    timespan: '24h',
  });

  const res = await fetch(`${GDELT_DOC_API}?${params}`);
  if (!res.ok) {
    throw new Error(`GDELT API error: ${res.status} ${res.statusText}`);
  }

  const data = (await res.json()) as { articles?: Array<{
    url: string;
    title: string;
    seendate: string;
    domain: string;
    language: string;
    sourcecountry: string;
  }> };

  const articles = (data.articles ?? []).map((a) => ({
    title: a.title,
    url: a.url,
    date: a.seendate,
    source: a.domain,
    country: a.sourcecountry,
  }));

  return {
    source: 'gdelt',
    fetchedAt: new Date().toISOString(),
    content: articles,
  };
}

File: augur-engine/src/collect/gdelt.py
"""GDELT Cloud API collector."""

from __future__ import annotations

from datetime import datetime, timezone

import httpx

from ..config.types import Signal

GDELT_DOC_API = "https://api.gdeltproject.org/api/v2/doc/doc"


async def collect_gdelt() -> Signal:
    params = {
        "query": "theme:ECON_BANKRUPTCY OR theme:ENV_CLIMATECHANGE OR theme:CRISISLEX_CRISISLEXREC OR theme:MILITARY",
        "mode": "ArtList",
        "maxrecords": "20",
        "format": "json",
        "sort": "DateDesc",
        "timespan": "24h",
    }

    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.get(GDELT_DOC_API, params=params)
        resp.raise_for_status()
        data = resp.json()

    articles = [{
        "title": a.get("title", ""),
        "url": a.get("url", ""),
        "date": a.get("seendate", ""),
        "source": a.get("domain", ""),
        "country": a.get("sourcecountry", ""),
    } for a in data.get("articles", [])]

    return Signal(
        source="gdelt",
        fetched_at=datetime.now(timezone.utc).isoformat(),
        content=articles,
    )
